core/frontend/src/test/tile/ImdlParser.test.ts

```diff
@@ -0,0 +1,276 @@
+    const m2 = acquire(RS.ModelTimeline.fromJSON(model2Props));
+    expect(m1).not.to.equal(m2);
+
+    const s1 = acquire(RS.Script.fromJSON(script1Props)!);
+    const s2 = acquire(RS.Script.fromJSON(script2Props)!);
+    expect(s1).not.to.equal(s2);
+
+    m1.release();
+    m2.release();
+    s1.release();
+    s2.release();
+  });
+
+  it("returns a new parser for the same timeline after original parser is released", () => {
+    const m = RS.ModelTimeline.fromJSON(model1Props);
+    const mp1 = acquire(m); // ref-count = 1
+    const mp2 = acquire(m); // ref-count = 2
+    expect(mp1).to.equal(mp2);
+
+    mp1.release(); // ref-count = 1
+    const mp3 = acquire(m); // ref-count = 2
+    expect(mp3).to.equal(mp1);
+
+    mp1.release(); // ref-count = 1
+    mp1.release(); // ref-count = 0
+    const mp4 = acquire(m);
+    expect(mp4).not.to.equal(mp1);
+    expect(acquire(m)).to.equal(mp4);
+
+    mp4.release();
+    mp4.release();
+
+    const s = RS.Script.fromJSON(script1Props)!;
+    const sp1 = acquire(s); // ref-count = 1
+    const sp2 = acquire(s); // ref-count = 2
+    expect(sp1).to.equal(sp2);
+
+    sp1.release(); // ref-count = 1
+    const sp3 = acquire(s); // ref-count = 2
+    expect(sp3).to.equal(sp1);
+
+    sp1.release(); // ref-count = 1
+    sp1.release(); // ref-count = 0
+    const sp4 = acquire(s);
+    expect(sp4).not.to.equal(sp1);
+    expect(acquire(s)).to.equal(sp4);
+
+    sp4.release();
+    sp4.release();
+  });
+
+  it("returns a different parser each time if not using web worker", () => {
+    const parsers = new Set<ImdlParser>();
+    const getParser = (timeline?: ImdlTimeline) => {
+      const parser = acquireImdlParser({ noWorker: true, timeline });
+      expect(parsers.has(parser)).to.be.false;
+      parsers.add(parser);
+    };
+
+    getParser();
+    getParser();
+    const model = RS.ModelTimeline.fromJSON(model1Props);
+    getParser(model);
+    getParser(model);
+    const script = RS.Script.fromJSON(script2Props);
+    getParser(script);
+    getParser(script);
+  });
+});
+
+describe("ImdlParser", () => {
+  before(async () => IModelApp.startup());
+  after(async () => IModelApp.shutdown());
+
+  it("produces an error upon invalid tile header", async () => {
+    const parser = acquireImdlParser({});
+    const document = await parser.parse({
+      data: new Uint8Array(512),
+      batchModelId: "0x123",
+      is3d: true,
+      maxVertexTableSize: 4096,
+    });
+
+    expect(document).to.equal(TileReadStatus.InvalidHeader);
+  });
+
+  it("produces an error when reading outside tile data", async () => {
+    const parser = acquireImdlParser({});
+    await expect(parser.parse({
+      data: new Uint8Array(12),
+      batchModelId: "0x123",
+      is3d: true,
+      maxVertexTableSize: 4096,
+    })).to.be.eventually.rejectedWith(RangeError);
+  });
+
+  it("transfers the array buffer to the worker and back", async () => {
```

Comment:
What does "and back" mean here, since data gets moved and you are checking that its length is 0?